### PR TITLE
Feature/filter datasets by country

### PR DESCRIPTION
**Pull request by edbrett**
Following the previous PR to create countries tags for the map menu, now it needs to be connected to the data to filter the datasets menu based on the countries available. This is mostly selectors refactoring and can be a little hard to follow. Before reviewing be aware that only the most blazing practices have been applied to make the selectors as blazing as possible 🔥 💨 . Therefore complex review of these is likely unneeded.

Key points that lead to these blazing levels of performance:
- refactor the `map-selectors` to parse all datasets initially allow the selector to be reused by other components.
- refactor the `menu-selectors` to use map datasets and parse in sections based on countries selected and categories from the config.
- display countries as subCategories above global layers when that country is selected.
- display sentence for countries that have no layers in that datasets tab.
- show countries statement in legend for layers that are non global and are available for certain countries "This layer is only available for certain countries" -> tooltip showing countries list.

BONUS (don't add to blazingness, but still 🔥):
- updated layer manager
- minor structure change to `<Pill />` component
- loader for when datasets on the menu are loading
- fix an old issue with the `<Threshold Selector />` in the Legend components from getting the correct Layer data.

### CODE REVIEW

**Comment by benlaken** [comment]
Looks like its working really well 👍

File: app/javascript/components/map/components/legend/legend-component.jsx [hunk 1]
```diff
@@ -58,10 +60,11 @@ class MapLegend extends Component {
...
                   isLossLayer,
                   metadata,
-                  id
+                  id,
+                  global,
+                  iso
                 } = lg;
-                const activeLayers = lg.layers.filter(l => l.active) || [];
-                const activeLayer = activeLayers && activeLayers[0];
+                const activeLayer = lg.layers.find(l => l.active) || [];
```
[the remark this answers is not in the document]
Comment (edbrett) [reply app/javascript/components/map/components/legend/legend-component.jsx]: this might have been addressed in a later PR, but correct!

File: app/javascript/components/map/components/legend/legend-component.jsx [hunk 2]
```diff
@@ -132,6 +135,13 @@ class MapLegend extends Component {
                     {isLossLayer && (
                       <LossStatement className="loss-statement" />
                     )}
+                    {global &&
+                      !isEmpty(iso) && (
```
[the remark this answers is not in the document]
Comment (edbrett) [reply app/javascript/components/map/components/legend/legend-component.jsx]: In this case, the API return an empty array. I can handle it by checking in the selector manually and then spreading back but it seems like just as much work to me.

File: app/javascript/components/ui/pill/component.jsx [hunk 3]
```diff
@@ -7,10 +7,14 @@ import closeIcon from 'assets/icons/close.svg';
...
 
 const Button = props => {
-  const { label, onRemove } = props;
+  const { label, onRemove, active } = props;
 
   return (
-    <div className={`c-pill ${onRemove ? '-removable' : ''}`}>
+    <div
+      className={`c-pill ${onRemove ? '-removable' : ''} ${
+        !active ? '-inactive' : ''
```
[the remark this answers is not in the document]
Comment (edbrett) [reply app/javascript/components/ui/pill/component.jsx]: This one is tricky. As it isnt going to be disabled, just a different color to represent no data, so you can still remove it. I will go with active.

File: app/javascript/pages/map/menu/components/sections/datasets/component.jsx [hunk 4]
```diff
@@ -66,6 +68,16 @@ class Datasets extends PureComponent {
             )}
           </div>
         </div>
+        {countriesWithoutData && (
+          <p
+            className="no-datasets-message"
+            dangerouslySetInnerHTML={{
+              __html: `No datasets available for <strong>${countriesWithoutData.join(
```
[the remark this answers is not in the document]
Comment (edbrett) [reply app/javascript/pages/map/menu/components/sections/datasets/component.jsx]: Because I didnt check this. Embarrasing.

File: app/javascript/pages/map/menu/menu-selectors.js [hunk 5]
```diff
@@ -132,16 +26,25 @@ export const getExploreSection = createSelector(
   settings => settings.exploreSection
 );
 
-const getUnselectedCountries = createSelector(
-  [getCountries, getMenuSettings],
+export const getAvailableCountries = createSelector(
+  [getCountries, getParsedDatasets],
+  (countries, datasets) => {
+    if (isEmpty(countries) || isEmpty(datasets)) return null;
+    const validIsos = flatten(datasets.filter(d => !d.global).map(d => d.iso));
```
[the remark this answers is not in the document]
Comment (edbrett) [reply app/javascript/pages/map/menu/menu-selectors.js]: In the API `d.iso` is never undefined. It always arrives as an array as part of the schema. `d.global` is something we set for the application. As a layer can have isos and be global and vice versa. I can still check for `d.iso` here though.

File: app/javascript/pages/map/menu/menu-selectors.js [hunk 6]
```diff
@@ -150,93 +53,123 @@ const getActiveCountries = createSelector(
...
-    };
-  })
+export const getSections = createSelector(
+  [getParsedDatasets, getActiveCountries],
+  (datasets, countries) => {
+    if (isEmpty(datasets)) return menuSections;
+    return menuSections.map(s => {
+      const { slug, subCategories } = s;
+      const sectionDatasets =
+        datasets && datasets.filter(d => d.tags.indexOf(slug) > -1);
```
[the remark this answers is not in the document]
Comment (edbrett) [reply app/javascript/pages/map/menu/menu-selectors.js]: Yes I have started using this but I old habits die hard.

File: app/javascript/pages/map/menu/menu-component.jsx [hunk 7]
```diff
@@ -84,6 +87,7 @@ class Menu extends PureComponent {
                 {...rest}
```
[the remark this answers is not in the document]
Comment (edbrett) [reply app/javascript/pages/map/menu/menu-component.jsx]: In this case, `activeSection` is created in the selectors, with specific keys for this component, so I am not worried about spreading it across the component. As the Component is a variable itself I passed these here to account for that. i.e. setMenuSettings is needed in whichever component is rendered, so I pass it. I am not sure I understand this part of the comment.

However, I think I will move `onToggleLayer` and `onInfoClick` to the toggle component itself as they always operate that way and it is bound to the map implicitly.

**Comment by edbrett** [comment]
@sorodrigo thanks for the comments. I will take a look at addressing them. `classnames` is a nice thought that we forgot about a long time ago. It would make the components more expressive. We can address the selectors styles at a later date and I will make a note.

I am not sure I agree on your `isEmpty` comment, as that is the purpose of the function. Agreed it means you can't see the type but I wouldn't say it is abusing it.

File: app/javascript/pages/map/menu/menu-component.jsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (sorodrigo) [reply app/javascript/pages/map/menu/menu-component.jsx]: I meant to say that `...rest` groups all props that you don't deconstruct. And that you might not need to extract `setMenuSettings` out of rest. But my bad I just noticed you're using it in line 77 also. 👍

File: app/javascript/components/map/components/legend/legend-component.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (edbrett) [reply app/javascript/components/map/components/legend/legend-component.jsx]: This is because it already defined above. I hate the use of 'layers'

File: app/javascript/components/map/map-selectors.js [hunk 8]
```diff
@@ -56,7 +56,7 @@ export const getMapSettings = createSelector(
   (urlState, widgetState) => ({
     ...initialState,
     ...urlState,
-    ...widgetState
+    ...(widgetState && widgetState)
```
[the remark this answers is not in the document]
Comment (edbrett) [reply app/javascript/components/map/map-selectors.js]: As I have rolled back the map to not use widgets just yet I am going to remove this dependancy anyway.

**Comment by edbrett** [comment]
@sorodrigo feel free to review again.

File: app/javascript/components/map/components/legend/legend-component.jsx [hunk 9]
```diff
@@ -58,13 +56,13 @@ class MapLegend extends Component {
...
-                  iso
-                } = lg;
-                const activeLayer = lg.layers.find(l => l.active) || [];
+                  statementConfig
+                } =
+                  lg || {};
+
+                const activeLayer = (lg && lg.layers.find(l => l.active)) || [];
                 const { legendConfig, params, timelineConfig } =
                   activeLayer || {};
```
Comment (sorodrigo) [inline app/javascript/components/map/components/legend/legend-component.jsx]: these two lines are incompatible either activeLayer is an object or an array
